### PR TITLE
Optimize storage of trivial data vectors

### PR DESCRIPTION
**Pull request by atuchin-m**
For https://github.com/brave/brave-browser/issues/21960
Another optimization of text_classifer model: 12mb => 6mb.
Before:
image
image

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: vendor/bat-native-ads/src/bat/ads/internal/ml/transformation/normalization_transformation_unittest.cc [hunk 1]
```diff
@@ -10,7 +10,6 @@
```
Comment (atuchin-m) [inline vendor/bat-native-ads/src/bat/ads/internal/ml/transformation/normalization_transformation_unittest.cc:10]: Surprisingly, this file already contains tests for both sparse/dense(trivial) cases. So looks like no extra tests are needed.
But I've added a few for `Normalize()`.

File: vendor/bat-native-ads/src/bat/ads/internal/ml/data/vector_data.h [hunk 2]
```diff
@@ -37,13 +40,12 @@ class VectorData final : public Data {
 
   void Normalize();
 
-  int GetDimensionCount() const;
+  int GetDimensionCountForTesting() const;
```
Comment (atuchin-m) [inline vendor/bat-native-ads/src/bat/ads/internal/ml/data/vector_data.h:43]: BTW, Something**ForTesting** means that you shouldn't use it outside the test code. I can be slow/limited/unstable/etc.
Chromium uses `PRESUBMIT.py` force this rule. I hope one day we will do the same.

https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md#tests-and-test_only-code

File: vendor/bat-native-ads/src/bat/ads/internal/ml/data/vector_data.cc [hunk 3]
```diff
@@ -9,79 +9,123 @@
...
+
 namespace ads {
 namespace ml {
 
 namespace {
 constexpr double kMinimumVectorLength = 1e-7;
 }  // namespace
 
-VectorData::VectorData() : Data(DataType::kVector) {}
+// A actual storage. Wrapped to a struct to make simple copy/move code.
```
Comment (iefremov) [inline vendor/bat-native-ads/src/bat/ads/internal/ml/data/vector_data.cc:21]: An actual storage

File: vendor/bat-native-ads/src/bat/ads/internal/ml/data/vector_data.cc [hunk 4]
```diff
@@ -9,79 +9,123 @@
...
+      : dimension_count_(dimension_count),
+        points_(std::move(points)),
+        values_(std::move(values)) {
+    DCHECK(points_.size() == values.size() || points_.empty());
+  }
+
+  size_t GetSize() const { return values_.size(); }
+
+  uint32_t GetPointAt(size_t index) const {
+    DCHECK_GE(index, 0u);
```
Comment (iefremov) [inline vendor/bat-native-ads/src/bat/ads/internal/ml/data/vector_data.cc:44]: always true for `size_t`

**Review by iefremov — APPROVED** [review]

File: vendor/bat-native-ads/src/bat/ads/internal/ml/data/vector_data.cc (hunk 4 above)
Comment (atuchin-m) [reply vendor/bat-native-ads/src/bat/ads/internal/ml/data/vector_data.cc]: Fixed

File: vendor/bat-native-ads/src/bat/ads/internal/ml/data/vector_data.cc (hunk 3 above)
Comment (atuchin-m) [reply vendor/bat-native-ads/src/bat/ads/internal/ml/data/vector_data.cc]: Fixed

File: vendor/bat-native-ads/src/bat/ads/internal/ml/data/vector_data.cc [hunk 5]
```diff
@@ -9,79 +9,123 @@
...
+    DCHECK_LT(index, values_.size());
+    if (points_.empty())  // The "dense" case, see the description.
+      return index;
```
[the remark this answers is not in the document]
Comment (atuchin-m) [reply vendor/bat-native-ads/src/bat/ads/internal/ml/data/vector_data.cc]: Fixed

**Review by aseren — APPROVED** [review]

**Pull request merged**